### PR TITLE
Add export products to .csv file feature

### PR DESCRIPTION
**Pull request by tanmaij**
Endpoint GET /api/v1/products/export-csv
Step 1: 
- Get and validate filter request from request body
- Call export function in service
Step 2: 
- Export function in service call toFilter function which is created in getProducts task(but pagination is empty)
- Call GetProducts function in repository with the filter object (old function)
Step 3:
- GetProducts function in repository will be changed something. If pagination is empty, it will query all record in database. Because pagination is set default value by GetProducts function in service now
- Service layer get data which is returned by repository layer, then service layer convert it to byte ( the file), then service save it to a .csv file and return file name 
- Handler layer get data from service layer. Response url with the file name

### CODE REVIEW

File: internal/handler/rest/v1/product.go [hunk 1]
```diff
@@ -311,3 +317,32 @@ func (h Handler) ImportProductCSV(w http.ResponseWriter, r *http.Request) {
...
+	// 3. Call export func and send mail
+	result, err := h.productServ.ExportProductsCSV(r.Context(), getProductsInput)
+	if err != nil {
+		handleProductError(w, err)
+		return
+	}
+
+	appURL := os.Getenv("APP_URL")
+	urlToDownload := appURL + "/files/" + result
+	utils.WriteJSONResponse(w, http.StatusOK, urlToDownload)
```
Comment (tanmaij) [inline internal/handler/rest/v1/product.go:347]: {"product_csv_url": "http://localhost:5000/products.csv"}

File: internal/service/product/product.go [hunk 2]
```diff
@@ -342,3 +351,38 @@ func (p impl) ImportProductCSV(ctx context.Context, fileName string, csvFile io.
...
+	fileBytes = append(fileBytes, headerRow...)
+
+	// Add the rows to the fileBytes
+	for _, p := range products {
+		row := []byte(fmt.Sprintf("%d,%s,%s,%f,%d,%v,%s,%v,%v\n", p.ID, p.Title, p.Description, p.Price, p.Quantity, p.IsActive, p.User.Name, p.CreatedAt, p.UpdatedAt))
+		fileBytes = append(fileBytes, row...)
+	}
+
+	// Create a file and check for errors
+	now := time.Now()
```
Comment (tanmaij) [inline internal/service/product/product.go:375]: Format EMD

File: internal/service/product/product.go [hunk 3]
```diff
@@ -342,3 +351,38 @@ func (p impl) ImportProductCSV(ctx context.Context, fileName string, csvFile io.
 
 	return nil
 }
+
+func (p impl) ExportProductsCSV(ctx context.Context, input GetProductsInput) (string, error) {
+	// Call repository to get filled product list
+	products, _, err := p.repo.Product().GetProducts(ctx, toFilter(input))
+	if err != nil {
+		return "", err
```
Comment (tanmaij) [inline internal/service/product/product.go:359]: Define ERR

File: internal/service/product/product.go [hunk 4]
```diff
@@ -342,3 +351,38 @@ func (p impl) ImportProductCSV(ctx context.Context, fileName string, csvFile io.
 
 	return nil
 }
+
+func (p impl) ExportProductsCSV(ctx context.Context, input GetProductsInput) (string, error) {
+	// Call repository to get filled product list
+	products, _, err := p.repo.Product().GetProducts(ctx, toFilter(input))
```
Comment (tanmaij) [inline internal/service/product/product.go:357]: TODO: handle big data case

Comment (tanmaij) [reply internal/service/product/product.go:357]: 1000

File: internal/handler/rest/v1/product.go (hunk 1 above)
Comment (tanmaij) [reply internal/handler/rest/v1/product.go:347]: DOne

File: internal/service/product/product.go (hunk 2 above)
Comment (tanmaij) [reply internal/service/product/product.go:375]: Done

File: internal/service/product/product.go (hunk 4 above)
Comment (tanmaij) [reply internal/service/product/product.go:357]: Done

File: internal/handler/rest/v1/product.go [hunk 5]
```diff
@@ -311,3 +319,67 @@ func (h Handler) ImportProductCSV(w http.ResponseWriter, r *http.Request) {
 
 	utils.WriteJSONResponse(w, http.StatusOK, fmt.Sprintf("Starting import product from csv: %s", fileHeader.Filename))
 }
+
+type exportProductsCSVRes struct {
```
Comment (tanmaij) [inline internal/handler/rest/v1/product.go:323]: productCSVResponse

File: internal/handler/rest/v1/product_test.go [hunk 6]
```diff
@@ -1028,3 +1018,215 @@ func TestProductHandler_ImportProductCSV(t *testing.T) {
...
+					UserID:     1,
+					OrderBy: productService.OrderInput{
+						Title:     "desc",
+						CreatedAt: "desc",
+					},
+				},
+				mockResultURL: "file_20222807.csv",
+			},
+			expOutput: output{
+				body:       exportProductsCSVRes{ProductCSVURL: "http://localhost:5000/api/v1/files/file_20222807.csv"},
```
Comment (tanmaij) [inline internal/handler/rest/v1/product_test.go:1063]: Define APP URL

File: internal/service/product/product.go [hunk 7]
```diff
@@ -342,3 +350,59 @@ func (p impl) ImportProductCSV(ctx context.Context, fileName string, csvFile io.
...
+	file, err := os.Create(filepath.Join(dir, filename))
+	if err != nil {
+		return "", ErrCannotCreateFile
+	}
+	defer file.Close()
+	// Write bytes to the file
+	file.Write(fileBytes)
+
+	//TODO: Send mail
+	go sendFileToMail(filename)
```
Comment (tanmaij) [inline internal/service/product/product.go:391]: Rename to sendEmail

File: internal/service/product/product.go [hunk 8]
```diff
@@ -342,3 +350,59 @@ func (p impl) ImportProductCSV(ctx context.Context, fileName string, csvFile io.
...
+
+	return filename, nil
+}
+
+func (p impl) GetCSVFile(ctx context.Context, filePath string) ([]byte, error) {
```
Comment (tanmaij) [inline internal/service/product/product.go:396]: DownloadCSV

File: internal/service/product/product.go [hunk 9]
```diff
@@ -342,3 +357,89 @@ func (p impl) ImportProductCSV(ctx context.Context, fileName string, csvFile io.
...
+	file, err := os.Create(filePath)
+	if err != nil {
+		return "", ErrCannotCreateFile
+	}
+	defer file.Close()
+	// Write bytes to the file
+	file.Write(fileBytes)
+
+	//TODO: Send mail
+	go sendEmail([]string{"dhuuloc8818@gmail.com"}, "Export products to CSV file", "<h1>Export product list to CSV file successfully!</h1>", []string{filePath})
```
Comment (vinhnv1) [inline internal/service/product/product.go:400]: You should create an input struct for this

File: internal/service/product/product.go [hunk 10]
```diff
@@ -342,3 +357,89 @@ func (p impl) ImportProductCSV(ctx context.Context, fileName string, csvFile io.
...
+		_, err := e.AttachFile(v)
+		if err != nil {
+			log.Println("Can't attach file")
+			log.Println(err)
+			return
+		}
+	}
+
+	// Authenticate and send email
+	auth := smtp.PlainAuth("", mail, pass, "smtp.gmail.com")
```
Comment (vinhnv1) [inline internal/service/product/product.go:436]: Define this host as an invironment variable

File: .env.dev [hunk 11]
```diff
@@ -1 +1,5 @@
 DB_URL=postgres://s3corp-golang-fresher:@localhost:5432/s3corp-golang-fresher?sslmode=disable
+APP_URL=http://localhost:5000
+MAIL=killdogvinam@gmail.com
+PASS=ybtkkhcjqcqxpitd
```
Comment (vinhnv1) [inline .env.dev:4]:
```suggestion
MAIL_USERNAME=killdogvinam@gmail.com
MAIL_PASS=ybtkkhcjqcqxpitd
```

**Pull request merged**